R/data.table.R

```diff
@@ -292,119 +301,99 @@ replace_dot_alias = function(e) {
-      # isub is a single symbol name such as B in DT[B]
-      i = try(eval(isub, parent.frame(), parent.frame()), silent=TRUE)
-      if (inherits(i,"try-error")) {
-        # must be "not found" since isub is a mere symbol
-        col = try(eval(isub, x), silent=TRUE)  # is it a column name?
-        msg = if (inherits(col,"try-error")) " and it is not a column name either."
-        else paste0(" but it is a column of type ", typeof(col),". If you wish to select rows where that column contains TRUE",
-                    ", or perhaps that column contains row numbers of itself to select, try DT[(col)], DT[DT$col], or DT[col==TRUE] is particularly clear and is optimized.")
-        stop(as.character(isub), " is not found in calling scope", msg,
-             " When the first argument inside DT[...] is a single symbol (e.g. DT[var]), data.table looks for var in calling scope.")
+        # #932 related so that !(v1 == 1) becomes v1 == 1 instead of (v1 == 1) after removing "!"
+        if (isub %iscall% "(" && !is.name(isub[[2L]]))
+          isub = isub[[2L]]
+      }
+      
+      if (!is.language(isub)) {
+        i = isub
+      } else if (length(o <- .prepareFastSubset(isub = isub, x = x,
+                                         enclos =  parent.frame(),
+                                         notjoin = notjoin, verbose = verbose))){
+        ## redirect to the is.data.table(x) == TRUE branch.
+        ## Additional flag to adapt things after bmerge:
+        optimizedSubset = TRUE
+        notjoin = o$notjoin
+        i = o$i
+        on = o$on
+        ## the following two are ignored if i is not a data.table.
+        ## Since we are converting i to data.table, it is important to set them properly.
+        nomatch = 0L
+        mult = "all"
+      }
+      else if (!is.name(isub)) {
+        ienv = new.env(parent=parent.frame())
+        if (getOption("datatable.optimize")>=1L) assign("order", forder, ienv)
+        i = tryCatch(eval(.massagei(isub), c(x, .N = nrow(x)), ienv), error=function(e) {
+          if (grepl(":=.*defined for use in j.*only", e$message))
+            stop("Operator := detected in i, the first argument inside DT[...], but is only valid in the second argument, j. Most often, this happens when forgetting the first comma (e.g. DT[newvar := 5] instead of DT[ , new_var := 5]). Please double-check the syntax. Run traceback(), and debugger() to get a line number.")
+          else
+            .checkTypos(e, names_x)
+        })
+      } else {
+        # isub is a single symbol name such as B in DT[B]
+        if (isub == quote(.N))
+          i = nrow(x)
+        else if (exists(as.character(isub), parent.frame()))
+          i = eval(isub, parent.frame(), parent.frame())
+        else {
+          # must be "not found" since isub is a mere symbol
+          col = try(eval(isub, x), silent=TRUE)  # is it a column name?
+          msg = if (inherits(col,"try-error")) " and it is not a column name either."
```

Comment:
this part needs to be piped through `gettext` / `gettextf` for potential translation